dlib/control/sarsa_abstract.h

```diff
@@ -0,0 +1,156 @@
+                - get_discount() >= 0 && get_discount() <= 1
+        !*/
+
+        void be_verbose (
+        );
+        /*!
+            ensures
+                - This object will print status messages to standard out so that a
+                  user can observe the progress of the algorithm.
+        !*/
+
+        void be_quiet (
+        );
+        /*!
+            ensures
+                - this object will not print anything to standard out
+        !*/
+
+        const feature_extractor& get_feature_extractor(
+        ) const;
+        /*!
+            ensures
+                - returns the feature extractor used by this object.
+        !*/
+
+        unsigned long get_max_iterations(
+        ) const;
+        /*!
+            ensures
+                - returns the maximum number of iterations of this object's training.
+        !*/
+
+        void set_max_iterations(
+            unsigned long value
+        );
+        /*!
+            requires
+                - value >= 0
+            ensures
+                - #get_max_iterations() == value
+        !*/
+
+        double get_learning_rate(
+        ) const;
+        /*!
+            ensures
+                - returns the learning rate used by this object.
+        !*/
+
+        void set_learning_rate(
+            double value
+        );
+        /*!
+            requires
+                - value >= 0 && value <= 1
+            ensures
+                - get_learning_rate() == value
+        !*/
+
+        double get_discount(
+        ) const;
+        /*!
+            ensures
+                - returns the discount value used by this object.
+        !*/
+
+        void set_discount(
+            double value
+        );
+        /*!
+            requires
+                - value >= 0 && value <= 1
+            ensures
+                - get_discount() == value
+        !*/
+
+        template <
+                typename vector_type
+                >
+        policy<feature_extractor> train(
+            const vector_type& samples
+        ) const;
+        /*!
+            requires
+                - samples.size() > 1
+                - samples is something with an interface that looks like
+                  std::vector<process_sample<feature_extractor>>.  That is, it should
+                  be some kind of array of process_sample objects.
+                - feature_extractor::force_last_weight_to_1 == false // TODO
+            ensures
+                - Trains a policy based on the given data and returns the results. The
+                  idea is to find a policy that will obtain the largest possible reward
+                  when run on the process that generated the samples. In particular,
+                  if the returned policy is P then:
+                    - P(S) == the best action to take when in state S.
+            notice
+                - Note that SARSA needs the actual next action taken so it doesn't process
```

Comment:
This should be clarified.  In particular, it matters that the training samples are given in the order they happened in the environment.  That should be very clearly explained.  Other things, like lspi don't have this requirement.  For lspi, all the samples could be randomly sampled and it would work fine.   

Really for this you might be better off with a train that takes a vector of vector of samples, where each inner vector is one trial.